### PR TITLE
Add chat format support dataset in SFT

### PR DESCRIPTION
**Pull request by libinta**
# What does this PR do?
This PR is to add conversation format dataset support for supervised finetune.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by libinta** [comment]
@sywangyi can you check again

***** train metrics *****
  epoch                       =      3.6364
  max_memory_allocated (GB)   =       94.11
  memory_allocated (GB)       =       63.23
  total_flos                  = 414294080GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.6435
  train_runtime               =  0:05:52.07
  train_samples_per_second    =       8.846
  train_steps_per_second      =       0.069
^M  0%|          | 0/3 [00:00<?, ?it/s]^M100%|██████████| 3/3 [00:01<00:00,  2.65it/s]
***** eval metrics *****
  epoch                       =     3.6364
  eval_loss                   =     1.6226
  eval_runtime                = 0:00:07.13
  eval_samples                =        751
  eval_samples_per_second     =     -5.742
  eval_steps_per_second       =      -0.28
  max_memory_allocated (GB)   =      94.11
  memory_allocated (GB)       =      67.82
  perplexity                  =     5.0664
  total_memory_available (GB) =      94.62
libin debug  trl-sft-chat <class 'list'>
PASSED
***** train metrics *****
  epoch                       =      7.2727
  max_memory_allocated (GB)   =       93.86
  memory_allocated (GB)       =       25.34
  total_flos                  = 828883840GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.9224
  train_runtime               =  0:06:41.86
  train_samples_per_second    =     123.449
  train_steps_per_second      =        0.06
^M  0%|          | 0/2 [00:00<?, ?it/s]^M100%|██████████| 2/2 [00:00<00:00,  2.54it/s]
***** eval metrics *****
  epoch                       =     7.2727
  eval_loss                   =     1.9267
  eval_runtime                = 0:00:14.74
  eval_samples                =        751
  eval_samples_per_second     =     -65.64
  eval_steps_per_second       =     -0.203
  max_memory_allocated (GB)   =      93.86
  memory_allocated (GB)       =       34.5
  perplexity                  =     6.8666
  total_memory_available (GB) =      94.62
libin debug  trl-sft-chat-peft <class 'list'>
PASSED

======================== 2 passed in 897.10s (0:14:57) =========================

File: examples/trl/README.md [hunk 1]
```diff
@@ -19,10 +51,12 @@ There are two main steps to the DPO training process:
     ```
     python ../gaudi_spawn.py --world_size 8 --use_mpi sft.py \
         --model_name_or_path meta-llama/Llama-2-7b-hf \
+        --dataset_name "lvwerra/stack-exchange-paired" \
         --output_dir="./sft" \
```
Comment (skaulintel) [inline examples/trl/README.md:55]: shouldn't we add streaming=False here as well?

File: tests/utils.py [hunk 2]
```diff
@@ -60,6 +60,7 @@
     "llama_guard": [("meta-llama/LlamaGuard-7b", "Habana/llama")],
     "code_llama": [("codellama/CodeLlama-13b-Instruct-hf", "Habana/llama")],
     "protst": [("mila-intel/protst-esm1b-for-sequential-classification", "Habana/gpt2")],
+    "qwen2": [("Qwen/Qwen2-72B", "Habana/qwen")],
```
Comment (sywangyi) [inline tests/utils.py:63]: should be Qwen2-7B?

**Comment by libinta** [comment]
@Wang, Yi ***@***.***> Can you check again? Thanks

**Comment by sywangyi** [comment]
pytest tests/test_examples.py::MultiCardSFTExampleTester fail
1)seq_length should be changed to max_seq_length in tests/baselines/llama_7b.json for sft test
2)streaming mode"True" + do_eval "True" coredump "TypeError: object of type 'IterableDataset' has no len()"

File: examples/trl/sft.py [hunk 3]
```diff
@@ -56,131 +73,151 @@ class ScriptArguments:
...
 
-def create_datasets(tokenizer, args, seed=None):
-    dataset = load_dataset(
-        args.dataset_name,
-        data_dir=args.subset,
-        split=args.split,
+    base_model = AutoModelForCausalLM.from_pretrained(
+        script_args.model_name_or_path,
+        low_cpu_mem_usage=low_cpu_mem_usage,
+        torch_dtype=torch.bfloat16,
```
Comment (mfuntowicz) [inline examples/trl/sft.py:160]: Checking on this @libinta : Do you want/need to force `torch.bfloat16` for the model?

**Review by mfuntowicz — APPROVED** [review]

**Review by regisss — COMMENTED** [review]
There is a merge conflict to solve

File: examples/trl/sft.py [hunk 4]
```diff
@@ -56,131 +73,151 @@ class ScriptArguments:
...
+            dataset = load_dataset(
+                args.dataset_name,
+                data_dir=args.subset,
+                split=args.split,
+                token=script_args.token,
+                num_proc=args.num_workers if not args.streaming else None,
+                streaming=args.streaming,
+            )
+        elif args.dataset_path:
+            ext = args.dataset_path.split(".")[-1]
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/trl/sft.py:127]: if user provides the local data file

File: examples/trl/README.md [hunk 5]
```diff
@@ -7,6 +7,38 @@ First, you should install the requirements:
 ```
 $ pip install -U -r requirements.txt
 ```
+## Supervised Finetune
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/trl/README.md:10]: Done

File: examples/trl/sft.py [hunk 6]
```diff
@@ -56,131 +73,151 @@ class ScriptArguments:
...
+            train_data = dataset.skip(args.size_valid_set)
+            train_data = train_data.shuffle(buffer_size=args.shuffle_buffer, seed=seed)
+        else:
+            dataset = dataset.train_test_split(test_size=args.validation_split_percentage * 0.01, seed=seed)
+            train_data = dataset["train"]
+            valid_data = dataset["test"]
+            logger.info(f"Size of the train set: {len(train_data)}. Size of the validation set: {len(valid_data)}")
+        if args.dataset_name == "lvwerra/stack-exchange-paired":
+            chars_per_token = chars_token_ratio(train_data, tokenizer)
+            logger.info(f"The character to token ratio of the dataset is: {chars_per_token:.2f}")
```
[the remark this answers is not in the document]
Comment (libinta) [reply examples/trl/sft.py:143]: it for question/answer format as the stack-exchange-paired

File: examples/trl/sft.py (hunk 3 above)
Comment (libinta) [reply examples/trl/sft.py:159]: yes

File: examples/trl/README.md [hunk 7]
```diff
@@ -7,6 +7,37 @@ First, you should install the requirements:
 ```
 $ pip install -U -r requirements.txt
 ```
+## Supervised Finetuneing
```
Comment (regisss) [inline examples/trl/README.md:10]:
```suggestion
## Supervised Finetuning
```

File: examples/trl/sft.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (regisss) [reply examples/trl/sft.py:127]: Why giving the extension as the path to the dataset? I don't understand that, here is the doc: https://huggingface.co/docs/datasets/package_reference/loading_methods#datasets.load_dataset

**Review by regisss — APPROVED** [review]

**Pull request merged**